wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/model/CheServiceImpl.java

```diff
@@ -0,0 +1,485 @@
+     *     <li>service_name:ro</li>
+     *     <li>service_name:rw</li>
+     * </ul>
+     */
+    public List<String> getVolumesFrom() {
+        if (volumesFrom == null) {
+            volumesFrom = new ArrayList<>();
+        }
+        return volumesFrom;
+    }
+
+    public void setVolumesFrom(List<String> volumesFrom) {
+        this.volumesFrom = volumesFrom;
+    }
+
+    public CheServiceImpl withVolumesFrom(List<String> volumesFrom) {
+        this.volumesFrom = volumesFrom;
+        return this;
+    }
+
+    /**
+     * Memory limit for the container of service, specified in bytes.
+     */
+    public Long getMemLimit() {
+        return memLimit;
+    }
+
+    public void setMemLimit(Long memLimit) {
+        this.memLimit = memLimit;
+    }
+
+    public CheServiceImpl withMemLimit(Long memLimit) {
+        this.memLimit = memLimit;
+        return this;
+    }
+
+    /**
+     * List of networks that should be connected to service.
+     */
+    public List<String> getNetworks() {
+        if (networks == null) {
+            networks = new ArrayList<>();
+        }
+        return networks;
+    }
+
+    public void setNetworks(List<String> networks) {
+        this.networks = networks;
+    }
+
+    public CheServiceImpl withNetworks(List<String> networks) {
+        this.networks = networks;
+        return this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof CheServiceImpl)) return false;
+        CheServiceImpl service = (CheServiceImpl)o;
+        return Objects.equals(id, service.id) &&
+               Objects.equals(containerName, service.containerName) &&
+               Objects.equals(command, service.command) &&
+               Objects.equals(entrypoint, service.entrypoint) &&
+               Objects.equals(image, service.image) &&
+               Objects.equals(dependsOn, service.dependsOn) &&
+               Objects.equals(environment, service.environment) &&
+               Objects.equals(expose, service.expose) &&
+               Objects.equals(ports, service.ports) &&
+               Objects.equals(labels, service.labels) &&
+               Objects.equals(links, service.links) &&
+               Objects.equals(volumes, service.volumes) &&
+               Objects.equals(volumesFrom, service.volumesFrom) &&
+               Objects.equals(memLimit, service.memLimit) &&
+               Objects.equals(build, service.build) &&
+               Objects.equals(networks, service.networks);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects
+                .hash(id,
+                      containerName,
+                      command,
+                      entrypoint,
+                      image,
+                      dependsOn,
+                      environment,
+                      expose,
+                      ports,
+                      labels,
+                      links,
```

Comment:
This may be wrong for 2 equal instances, consider the following example:

```java
CheServiceImpl s1 = new CheServiceImpl();
CheServiceImpl s2 = new CheServiceImpl();
s1.getLinks();
assert sq.equals(s2); // <- this will fail.
```
Explanation: an empty `ArrayList#hashCode` returns 1, while `Object.hashCode(null)` returns 0
:cry: